src/apps/socket/raw.lua

```diff
@@ -1,72 +1,123 @@
-module(...,package.seeall)
+module(..., package.seeall)
 
-local app    = require("core.app")
-local link   = require("core.link")
+local S = require("syscall")
+local h = require("syscall.helpers")
+local bit = require("bit")
+local link = require("core.link")
 local packet = require("core.packet")
-local dev    = require("apps.socket.dev").dev
+local ffi = require("ffi")
+local C = ffi.C
+
+local c, t = S.c, S.types.t
 
 RawSocket = {}
 
 function RawSocket:new (ifname)
    assert(ifname)
-   self.__index = self
-   return setmetatable({dev = dev:new(ifname)}, self)
+   local tp = h.htons(c.ETH_P["ALL"])
+   local sock, err = S.socket(c.AF.PACKET, bit.bor(c.SOCK.RAW, c.SOCK.NONBLOCK), tp)
+   if not sock then return nil, err end
+   local index, err = S.util.if_nametoindex(ifname, sock)
+   if err then
+      S.close(sock)
+      return nil, err
```

Comment:
The engine does not support this `return nil, err` return value. Rather it is always expecting an app instance to be returned. I think the best action here would be to raise an exception with `error(...reason...)` which the engine then should catch and handle (doesn't right now by default - that is something else we need to fix separately).